File: sample_app/lib/data/llm/yandexgpt_usecase.dart
import 'dart:convert';

import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:http/http.dart' as http;
import 'package:sample_app/domain/llm_usecase.dart';
import 'package:sample_app/models/app_settings.dart';

/// Реализация юзкейса для YandexGPT Chat Completions API
/// Документация: https://yandex.cloud/ru/docs/foundation-models/quickstart/yandexgpt
class YandexGptUseCase implements LlmUseCase {
  // Базовый URL берем из .env: YANDEX_GPT_BASE_URL
  // Фоллбэк — официальный endpoint.
  String get _endpoint => dotenv.env['YANDEX_GPT_BASE_URL'] ??
      'https://llm.api.cloud.yandex.net/foundationModels/v1/completion';

  String get _iamToken =>
      dotenv.env['YANDEX_IAM_TOKEN'] ?? dotenv.env['YC_IAM_TOKEN'] ?? '';
  String get _apiKey => dotenv.env['YANDEX_API_KEY'] ?? '';
  String get _folderId => dotenv.env['YANDEX_FOLDER_ID'] ?? '';

  String get _modelUri {
    // Позволяем переопределить модель через .env (YANDEX_MODEL_URI), иначе используем дефолт
    final override = dotenv.env['YANDEX_MODEL_URI'];
    if (override != null && override.isNotEmpty) return override;
    // По умолчанию используем yandexgpt/latest; можно заменить на yandexgpt-lite/latest
    final folder = _folderId;
    return 'gpt://$folder/yandexgpt';
  }

  @override
  Future<String> complete({
    required List<Map<String, String>> messages,
    required AppSettings settings,
  }) async {
    // Требуем либо IAM токен (предпочтительно), либо Api-Key (временный fallback)
    if (_iamToken.isEmpty && _apiKey.isEmpty) {
      throw Exception('Не найден Yandex IAM токен или API ключ. Укажите YANDEX_IAM_TOKEN (предпочтительно) или YANDEX_API_KEY в assets/.env');
    }
    if (_folderId.isEmpty) {
      throw Exception('Не указан YANDEX_FOLDER_ID в assets/.env');
    }

    // Конвертируем наше сообщение из {'role': 'user', 'content': '...'} в формат Yandex {'role': 'user', 'text': '...'}
    final ycMessages = [
      for (final m in messages)
        {
          'role': m['role'],
          'text': m['content'],
        }
    ];

    // Настройки генерации
    final double temperature = settings.yandexTemperature.clamp(0.0, 1.0);
    final int maxTokens = settings.yandexMaxTokens > 0 ? settings.yandexMaxTokens : 1;

    final body = jsonEncode({
      'modelUri': _modelUri,
      'completionOptions': {
        'stream': false,
        'temperature': temperature,
        'maxTokens': maxTokens,
      },
      'messages': ycMessages,
    });

    final headers = <String, String>{
      'Content-Type': 'application/json',
      if (_iamToken.isNotEmpty) ...{
        'Authorization': 'Bearer $_iamToken',
      } else ...{
        // Временный fallback на Api-Key (менее предпочтительно)
        'Authorization': 'Api-Key $_apiKey',
        'x-folder-id': _folderId,
      }
    };

    final response = await http.post(
      Uri.parse(_endpoint),
      headers: headers,
      body: body,
    );

    if (response.statusCode != 200) {
      throw Exception('Ошибка YandexGPT: ${response.statusCode} ${response.body}');
    }

    final Map<String, dynamic> data = jsonDecode(response.body);
    // Формат: { result: { alternatives: [ { message: { role, text } } ] } }
    final text = data['result']?['alternatives']?[0]?['message']?['text'];
    if (text is String && text.isNotEmpty) {
      return text;
    }
    throw Exception('Пустой ответ от YandexGPT');
  }
}

File: telegram_monitoring_agent/src/yandexgpt_usecase.py
#!/usr/bin/env python3
"""
YandexGPT LLM Implementation
"""

import aiohttp
import json
import os
from typing import List, Dict, Any
from .llm_usecase import LlmUseCase


class YandexGptUseCase(LlmUseCase):
    """YandexGPT LLM implementation"""

    def __init__(self):
        self.api_url = os.getenv('YANDEX_GPT_BASE_URL',
                                'https://llm.api.cloud.yandex.net/foundationModels/v1/completion')
        self.iam_token = os.getenv('YANDEX_IAM_TOKEN', '')
        self.api_key = os.getenv('YANDEX_API_KEY', '')
        self.folder_id = os.getenv('YANDEX_FOLDER_ID', '')
        self.model_uri = os.getenv('YANDEX_MODEL_URI', f'gpt://{self.folder_id}/yandexgpt')

    async def complete(self, messages: List[Dict[str, str]], **kwargs) -> str:
        """Complete the conversation using YandexGPT API"""

        # Check authentication
        if not self.iam_token and not self.api_key:
            raise Exception("YandexGPT authentication not found. Set YANDEX_IAM_TOKEN or YANDEX_API_KEY")

        if not self.folder_id:
            raise Exception("YandexGPT folder ID not found. Set YANDEX_FOLDER_ID")

        # Get parameters from kwargs or use defaults
        temperature = kwargs.get('temperature', 0.7)
        max_tokens = kwargs.get('max_tokens', 2000)

        # Convert messages to YandexGPT format
        yc_messages = []
        for msg in messages:
            yc_messages.append({
                'role': msg['role'],
                'text': msg['content']
            })

        request_body = {
            'modelUri': self.model_uri,
            'completionOptions': {
                'stream': False,
                'temperature': temperature,
                'maxTokens': max_tokens,
            },
            'messages': yc_messages,
        }

        headers = {'Content-Type': 'application/json'}

        if self.iam_token:
            headers['Authorization'] = f'Bearer {self.iam_token}'
        else:
            headers['Authorization'] = f'Api-Key {self.api_key}'
            headers['x-folder-id'] = self.folder_id

        async with aiohttp.ClientSession() as session:
            async with session.post(self.api_url, headers=headers, json=request_body) as response:
                if response.status != 200:
                    error_text = await response.text()
                    raise Exception(f"YandexGPT API error {response.status}: {error_text}")

                data = await response.json()
                text = data.get('result', {}).get('alternatives', [{}])[0].get('message', {}).get('text', '')

                if not text:
                    raise Exception("Empty response from YandexGPT")

                return text
